### PR TITLE
Align interactions when plot created inside scaled and/or translated elements using CSS transform

### LINKED ISSUE #888: bug: interactions misaligned when plot is inside transform: scale()-ed element
See http://codepen.io/hellochar/pen/JKgXPE and mouse around.

Looks like plotly.js miscalculates where the mouse is positioned if the plot itself is rendered inside an element with a `transform: scale()` CSS property set. The tooltips don't match up with where the mouse is. This is most apparent when you try to click+drag an area of the plot and it clearly shows the locations don't match up.

The buttons in the toolbar line up correctly, as does the hitbox for dragging the axes or enabling/disabling specific traces.

Also it looks like if you're at the top-left corner of the plot, the mouse position is accurate (that is, the (0, 0) origin point is handled correctly; only the scaling factor is off).

Also see https://github.com/palantir/plottable/issues/1644 which looks to be the same issue in Plottable.

Let me know if that makes sense and if there's something else I can do!

### PR DESCRIPTION
**Pull request by alexhartstone**
Fixes #888.
This is the first commit of a few fixes for interactions with plots that are broken due when css transform scaling is applied. 
The general aim is to undo the transform matrices applied to these elements for the interactions.

TODOs:
- [x] clicking the end of an axis in 2D cartesian to type a new number: the text entry box shows up in the wrong place
- [x] fixing 2D cartesian `pan`
- [x] fixing `zoom` & `pan` in `ternary` & `polar`
- [x] fixing `scattermapbox` `hover` & `select`
- [x] fixing `choroplethmapbox` `hover`
- [x] fixing the position of `hover` on `sankey` trace
- [x] fixing the position of `hover` on `parcats` trace
- [x] adding `hover`, `select` and `drag-zoom` tests for `cartesian` subplots
- [x] adding `hover`, `select` and `drag-zoom` tests for `polar` subplots
- [x] adding `hover`, `select` and `drag-zoom` tests for `mapbox` subplots i.e. `choroplethmapbox` and `scattermapbox`
- [x] adding `hover`, `select` and `drag-zoom` tests for `geo` subplots e.g. `choropleth`
- [x] adding `hover`, `select` and `drag-zoom` tests for `ternary` subplots
- [x] test using different browsers namely Firefox, Chrome and Safari

### CODE REVIEW

**Comment by archmoj** [comment]
Thanks very much @alexhartstone for submitting this PR.
I spend some time testing this on different plot types.
Please view/pick my commits on this dev branch: https://github.com/plotly/plotly.js/compare/pull-5193_ms that fixes few minor issues to start.

**Comment by archmoj** [comment]
@alexhartstone FYI - I also added few more commits to that branch that may be of interest.

**Comment by archmoj** [comment]
@alexhartstone just to let you know. I am investing more time on the dev-branch today. I will let you when I was able to complete my journey.

**Comment by archmoj** [comment]
@alexhartstone please pick my commits from https://github.com/plotly/plotly.js/compare/pull-5193_ms or alternatively allow us as the maintainers access to push your branch. Please find the instructions here https://github.com/plotly/plotly.js/blob/master/.github/PULL_REQUEST_TEMPLATE.md

With all those fixes the transforms e.g. `scale`, `scaleX`, `scaleY`, `translate` now works as expected on `cartesian`, `geo`, `gl2d`, `gl3d`, `ternary` and `polar` plots.
You should find a similar fix for the `mapbox` plot.
Also `gl3d` hovers does not scale according to the transform at the moment. It might be a desirable behaviour actually (and maybe we should try achieve the same with other subplots). However you should investigate why this is happening (if related to `pixelRatio` or not, etc).
Then you should start adding a number of tests for different subplots to ensure the `hover`, `select` and `zoom` works as expected.
Also it would be nice to catch the positioning issue when `rotation` is  present. 
Thanks!

**Comment by alexhartstone** [comment]
@archmoj thanks for this! I was looking at this as part of one of my sprints - I'll give you guys access to this branch.

**Comment by alexhartstone** [comment]
@archmoj  Thanks for this! I'm trying to allow edits for you but I'm having some trouble as I can't see the allow edits checkbox - looking into it

**Comment by archmoj** [comment]
> @archmoj Thanks for this! I'm trying to allow edits for you but I'm having some trouble as I can't see the allow edits checkbox - looking into it

@alexhartstone Here is more info: https://docs.github.com/en/free-pro-team@latest/github/collaborating-with-issues-and-pull-requests/allowing-changes-to-a-pull-request-branch-created-from-a-fork#enabling-repository-maintainer-permissions-on-existing-pull-requests

**Comment by archmoj** [comment]
Or alternatively you could search for how to `cherry-pick` from `another repository`.

**Comment by archmoj** [comment]
@alexhartstone FYI - I added few more commits for you to [my dev-branch](https://github.com/plotly/plotly.js/compare/pull-5193_ms) on top of your work including fixes for `gl3d` and optimizations for performance, etc.

Here are the remaining items I will pass to you so that we could merge this thing.
- [ ] fixing `mapbox` hover
- [ ] adding `hover` test for various subplots
- [ ] adding `select` test for various subplots
- [ ] adding `drag` test for various subplots
- [ ] possibly fixing positions when `rotation` is present (please note: this could be addressed in a separate PR later)

**Comment by archmoj** [comment]
@alexhartstone also please consider https://github.com/plotly/plotly.js/commit/2c810fcf066d3f706097427942888de8c5780b8e to handle css transform matrices of length 16.

**Comment by alexhartstone** [comment]
@archmoj hey mate, sorry I took a bit to get back to you. I switched from working on this task to another - I'm talking to my supervisor, hopefully can get back onto working on this as I would love to keep collaborating. I've tried inviting you to this fork so you can edit, as I still cannot see any checkbox for allowing edits - did it work?

**Comment by archmoj** [comment]
> @archmoj hey mate, sorry I took a bit to get back to you. I switched from working on this task to another - I'm talking to my supervisor, hopefully can get back onto working on this as I would love to keep collaborating. I've tried inviting you to this fork so you can edit, as I still cannot see any checkbox for allowing edits - did it work?

Yes. It worked. I was able to push those commits to you branch.
I think this PR is pretty close to a finish line.
When you could possibly add some tests to lock the current behaviour.

**Comment by alexhartstone** [comment]
@archmoj I'll let you guys know as soon as I get the go-ahead from my manager :)

**Comment by alexhartstone** [comment]
@archmoj just quickly looking through the commits - the hover event fix (and potentially in other places) now uses an inverse transform matrix attached to fullLayout (cf hover.js, line 339) that gets set once on plot initialisation. In our use case, transforms are applied dynamically, which is why I had originally done the transformation on each call to _hover - thus, this fix no longer works in our case, because the inverse transform matrix is for a transform that may not be valid.

Is it too slow to do it in every call? If so, what other options are there? Is there a way to update the fullLayout's inverseTransform when scaling is applied, or mark a plot as dynamic?

**Comment by archmoj** [comment]
> Is it too slow to do it in every call? If so, what other options are there? Is there a way to update the fullLayout's inverseTransform when scaling is applied, or mark a plot as dynamic?

Not way too slow. But it might be something we could enable via an option? @alexhartstone @alexcjohnson

File: src/lib/dom.js
Comment (alexcjohnson) [inline src/lib/dom.js:130]: Do you always get the `matrix` form? There are lots of other options listed in https://developer.mozilla.org/en-US/docs/Web/CSS/transform

Minor: `.map(function(n) {return +n;})` can be `.map(Number)`

**Comment by alexcjohnson** [comment]
@archmoj is there a codepen around where I can test this out? It looks like this approach will not support rotation, though I'm not sure there's any way around that given our heavy use of `getBoundingClientRect`. But I'm wondering if we need to apply this to all the uses of `getBoundingClientRect`...

File: src/lib/dom.js
Comment (alexhartstone) [reply src/lib/dom.js:130]: If you do something like element.style.transform, you can get any of the options listed there. But if you do it via window.getComputedStyle, I believe it always gives you a matrix of that form.

Comment (alexcjohnson) [reply src/lib/dom.js:130]: Good to know! Is that documented somewhere? If not we will want to test it in all supported browsers.

**Comment by alexhartstone** [comment]
@alexcjohnson @archmoj I have the go ahead from my end to start writing tests, so I'll get onto that. Writing one test for each interaction for the fixed chart subtypes - I'll put then into the interaction test file (eg click interactions into click.js jasmine test file), if that's not the best place let me know!

PS early on I did play around with fixing rotation but had very little success. I think it would take a larger restructuring of code - I think you can do it if you only use matrix transforms, but mixing in getBoundingClientRect with it makes it a bit tricky

File: src/lib/dom.js
Comment (alexhartstone) [reply src/lib/dom.js:130]: Hm, I've been unable to find a clear bit of documentation to confirm this, just lots of uses of it on stackoverflow. I think it's legit but it definitely needs testing.

**Comment by archmoj** [comment]
@alexhartstone FYI - I just fixed up the order of matrix so that `translate` works as well as `scale` https://github.com/plotly/plotly.js/pull/5193/commits/c409a28fd91d27ae4d845dc5f4d64919a420e36b.
Now it would be great if you add the tests to lock down the current behaviour for both `scale` and `translate`.
Thank you!

**Comment by archmoj** [comment]
> @archmoj is there a codepen around where I can test this out? It looks like this approach will not support rotation, though I'm not sure there's any way around that given our heavy use of `getBoundingClientRect`. But I'm wondering if we need to apply this to all the uses of `getBoundingClientRect`...

Here is the [codepen](https://codepen.io/MojtabaSamimi/pen/ZEOrPKO?editors=0100) to play with different CSS options and plot types.

File: test/jasmine/tests/cartesian_interact_test.js
Comment (archmoj) [inline test/jasmine/tests/cartesian_interact_test.js:2301]: Let's simply pass `transformString` to this function.
```
function transformPlot(gd, transformString) {
```
That way we could easily test various combinations of translates and scales.

File: test/jasmine/tests/cartesian_interact_test.js
Comment (archmoj) [inline test/jasmine/tests/cartesian_interact_test.js:2336]: Once the arguments of `transformPlot` was revised, we could add tests for cases such as:
```
transform: scale(0.5);
transform: scaleX(0.7); 
transform: scaleY(0.7); 
transform: translate(10%, 20%); 
transform: translate(100px, 200px);
transform: scale(0.5) translateY(-50%);
```

**Comment by alexcjohnson** [comment]
@archmoj thanks for the codepen. I added a few more subplot types: https://codepen.io/alexcjohnson/pen/JjKvQWK?editors=0010

Hover looks good in all of them that I've tried so far! :tada:

But there are still some issues with drag interactions:
- 2D cartesian panning is incorrectly scaled - ie with that 50% scaling the plot moves half as much as the mouse moves. Applies to 2D panning on the plot itself as well as single axis panning by dragging the middle of the axis area and single-axis or corner zoom by dragging the ends of the axis area or the corners of the plot.
- Also clicking the end of an axis in 2D cartesian to type a new number: the text entry box shows up in the wrong place.
- Ternary & polar: all the zoom and pan interactions are misaligned and mis-scaled. In polar you can also grab the outer ring to rotate the whole graph, or grab the end of the radial axis to rotate where it's displayed, and for these the center point you're rotating about is incorrect.

File: src/lib/dom.js
Comment (alexcjohnson) [reply src/lib/dom.js:130]: I've found https://www.w3.org/TR/css-transforms-1/#serialization-of-the-computed-value

> A <transform-list> for the computed value is serialized to one <matrix()> function ...

and https://zellwk.com/blog/css-translate-values-in-javascript/

> Browsers turn transform values into 2d or 3d matrices depending on what transformations you applied.

Still seems like we should test this in IE11 at least, but modern browsers seem to be consistent in this.

**Comment by alexcjohnson** [comment]
Here's a fun one: In Chrome, panning on a `geo` subplot works correctly, but Firefox gets the scaling wrong, like in 2D cartesian.

File: test/jasmine/tests/cartesian_interact_test.js
Comment (alexhartstone) [reply test/jasmine/tests/cartesian_interact_test.js:2301]: You're right, that's better

File: src/lib/matrix.js
Comment (archmoj) [inline src/lib/matrix.js:127]: Possibly we shouldn't try validating the matrix like this; which could sometimes be inaccurate.
I recall getting a `NaN` value in the first element of the matrix.
Instead of this, one should investigate why it happens at first place.

**Comment by archmoj** [comment]
> @archmoj thanks for the codepen. I added a few more subplot types: https://codepen.io/alexcjohnson/pen/JjKvQWK?editors=0010
> 
> Hover looks good in all of them that I've tried so far! 
> 
> But there are still some issues with drag interactions:
> 
> * 2D cartesian panning is incorrectly scaled - ie with that 50% scaling the plot moves half as much as the mouse moves. Applies to 2D panning on the plot itself as well as single axis panning by dragging the middle of the axis area and single-axis or corner zoom by dragging the ends of the axis area or the corners of the plot.
> * Also clicking the end of an axis in 2D cartesian to type a new number: the text entry box shows up in the wrong place.
> * Ternary & polar: all the zoom and pan interactions are misaligned and mis-scaled. In polar you can also grab the outer ring to rotate the whole graph, or grab the end of the radial axis to rotate where it's displayed, and for these the center point you're rotating about is incorrect.

Thanks @alexcjohnson for testing those interactions.
@alexhartstone please investigate fixing them. The `ternary` and `polar` fix might be quite similar to `cartesian` changes in `src/plots/cartesian/select.js`.
Thank you!

**Comment by alexhartstone** [comment]
Will also try to organise/abstract some of my boilerplate testing code

**Comment by alexhartstone** [comment]
@archmoj I'm also investigating ternary/polar issues and will continue tomorrow.

**Comment by archmoj** [comment]
@alexhartstone good to see the progress here.
Would you mind checking the completed tasks in the PR description?
Thanks.

File: src/plot_api/plot_api.js
Comment (archmoj) [inline src/plot_api/plot_api.js:3718]: Let's also compute and stash `._inverseScale(X|Y)` here and use that from other places.
```js
fullLayout._inverseScaleX = Math.sqrt(m[0][0] * m[0][0] + m[0][1] * m[0][1] + m[0][2] * m[0][2]);
fullLayout._inverseScaleY = Math.sqrt(m[1][0] * m[1][0] + m[1][1] * m[1][1] + m[1][2] * m[1][2]);
```
Thanks!

**Comment by CharsiCuba** [comment]
A similar issue is encountered if the css property zoom is used.

**Comment by archmoj** [comment]
@alexhartstone FYI - I just fixed <i>clicking the end of an axis in 2D cartesian to type a new number: the text entry box shows up in the wrong place</i> in https://github.com/plotly/plotly.js/pull/5193/commits/84ede3191af35a2fa6bf2e4dcb0ccc1ac1c78649.

**Comment by archmoj** [comment]
@alexhartstone I also fixed `choroplethmapbox` and removed `scattermapbox` hack in https://github.com/plotly/plotly.js/pull/5193/commits/4980e24fd517f39e86bd7671768d3b1c7b3d2d9e and checked more items on the PR description.

So if you could finish the tests, this PR would be ready for the final review.
Also FYI - we are planning to publish a new `plotly.js` minor in the following days and we could possibly include these improvements if completed.
Thanks!

**Comment by alexhartstone** [comment]
@archmoj it looks as though ternary panning is still not fixed - investigating

**Comment by alexhartstone** [comment]
@archmoj oh also, has there been any talk of allowing some kind of update call to recalculate inverse transforms on the layout? I ask because our use case (and I think many others') requires dynamic transformation.

**Comment by archmoj** [comment]
> @archmoj oh also, has there been any talk of allowing some kind of update call to recalculate inverse transforms on the layout? I ask because our use case (and I think many others') requires dynamic transformation.

Sure. I'll bring that up in the next internal meeting this week to possibly add a new layout attribute. BTW let's address that in a separate PR right after merging this.

**Comment by alexcjohnson** [comment]
> A similar issue is encountered if the css property zoom is used.

`zoom` could certainly be included, though it [doesn't seem to be fully supported by browsers](https://developer.mozilla.org/en-US/docs/Web/CSS/zoom) so we'll need to ensure we're reacting to what the browser is _actually_ rendering, not just the attribute the user provided. And it may take a bit of experimentation to figure out exactly how this combines with `transform` (whether on the same element or another in the hierarchy).

File: test/jasmine/tests/cartesian_interact_test.js
Comment (alexhartstone) [inline test/jasmine/tests/cartesian_interact_test.js:2424]: Ah I see - you're right, that's better than doing the recalculate function. Somehow it slipped my mind

**Comment by alexhartstone** [comment]
@archmoj am I correct in saying we also need to test panning? That is handled differently to zooming, if I recall. Currently working on ternary tests

**Comment by alexhartstone** [comment]
I've also discovered that ternary hover labels are positioned incorrectly

**Comment by archmoj** [comment]
> I've also discovered that ternary hover labels are positioned incorrectly EDIT: weirdly, this is only happening in Jasmine tests, and not in the devtools page

I added a number of tests for `ternary` in https://github.com/plotly/plotly.js/pull/5193/commits/29e772d2d18ce158cd4c5ba4171c1df2260b56fa as well as in https://github.com/plotly/plotly.js/pull/5193/commits/34845ecd80631c4e6996b09fd407408a32ce6c92 and now most of the boxes in the PR description are checked.

But yes please move forward and fix and lock down any remaining undesirable behaviour you noticed.
Thanks.

**Comment by archmoj** [comment]
@alexhartstone I just noticed that there is a problem with the position of hover on `sankey` trace.
Could you please fix that?

**Comment by archmoj** [comment]
@alexhartstone Also the `parcat` hover position is not correct which is added to the list on the PR description.

File: src/lib/matrix.js
Comment (archmoj) [reply src/lib/matrix.js]: Resolved by https://github.com/plotly/plotly.js/pull/5193/commits/28a917861d17ca923078b521612a08cc248bd23e.

**Comment by archmoj** [comment]
@alexcjohnson This PR is now ready for review.
FYI - @alexhartstone will address minor issues with the position of hover in `parcats` and `sankey` as well.

**Comment by archmoj** [comment]
> @alexcjohnson This PR is now ready for review.
> FYI - @alexhartstone will address minor issues with the position of hover in `parcats` and `sankey` as well.

Issues with `parcats` and `sankey` nodes were addressed in https://github.com/plotly/plotly.js/pull/5193/commits/a2f34c939714270610587833771b4066219b44dc and https://github.com/plotly/plotly.js/pull/5193/commits/9fe42b6a6dee721d73acfbbce541ab0ac1255e62.

So this looks ready to :dancer:

**Comment by archmoj** [comment]
Notable changes to `mapbox` as a result of this PR i.e. no longer modifying `event.clientX` and Y to display hover.
mapbox-fix

File: src/plot_api/plot_api.js
Comment (alexcjohnson) [inline src/plot_api/plot_api.js:3719]:
```suggestion
        var m = fullLayout._inverseTransform = Lib.inverseTransformMatrix(Lib.getFullTransformMatrix(gd));
```

File: src/plots/polar/polar.js
Comment (alexcjohnson) [inline src/plots/polar/polar.js:1203]: Would it work to just do this conversion in calculating `x1` and `y1`? It always makes things more confusing when a variable changes meaning, and I don't see `dx` and `dy` getting used after this.
```suggestion
        var x1 = x0 + dx * fullLayout._inverseScaleX;
        var y1 = y0 + dy * fullLayout._inverseScaleY;
```

File: tasks/test_syntax.js
Comment (alexcjohnson) [inline tasks/test_syntax.js:116]: Has this new behavior been tested in FF? Looks like this just used in mapbox. According to https://developer.mozilla.org/en-US/docs/Web/API/MouseEvent/offsetX perhaps this "not implemented in FF" is long long obsolete, but I'm not sure what the "inconsistent with others" comment was about.

Comment (archmoj) [reply tasks/test_syntax.js:116]: I tested it in FF. And it works.

**Review by alexcjohnson — APPROVED** [review]
Beautiful - nice work @archmoj and @alexhartstone! Fantastic tests.

💃  ~pending confirmation that mapbox works right in FF~ looks like you've confirmed 🎉

**Comment by archmoj** [comment]
@alexhartstone please open a new issue describing why recomputing transform matrix on events is required in your particular application with a demo like [this codepen](https://codepen.io/MojtabaSamimi/pen/XWKQwEO?editors=1000).
Thank you!

**Comment by alexhartstone** [comment]
Awesome @archmoj  @alexcjohnson  thanks guys!

**Comment by nicolaskruchten** [comment]
Thank you so much for taking the initiative on this @alexhartstone !

**Comment by nlevari** [comment]
@alexhartstone, Thanks you so much for this fix. I spent days trying to workaround,  unscaling and  rescaling.
Just updated to the last version and everything works.

**Comment by alexhartstone** [comment]
@nlevari :) :) thanks for the nice comment! Much credit is due to @archmoj too for his work

**Pull request merged**
